app/workers/digest_email_generation_worker.rb

```diff
@@ -2,51 +2,58 @@ class DigestEmailGenerationWorker < ApplicationWorker
   sidekiq_options queue: :email_generation_digest
 
   def perform(digest_run_subscriber_id)
-    email = nil
+    email_ids = []
+
     DigestRun.transaction do
       digest_run_subscriber = DigestRunSubscriber.lock.find(digest_run_subscriber_id)
-
       return if digest_run_subscriber.processed_at
 
       digest_items = DigestItemsQuery.call(
         digest_run_subscriber.subscriber,
         digest_run_subscriber.digest_run,
       )
 
-      email = create_email(digest_run_subscriber, digest_items)
+      email_ids = digest_items.map do |digest_item|
+        create_email(digest_run_subscriber, digest_item)
+      end
+
+      fill_subscription_content(
+        email_ids,
+        digest_items,
+        digest_run_subscriber,
+      )
+
       digest_run_subscriber.update!(processed_at: Time.zone.now)
     end
 
-    if email
-      SendEmailWorker.perform_async_in_queue(email.id, queue: :send_email_digest)
+    email_ids.each do |email_id|
+      SendEmailWorker.perform_async_in_queue(email_id, queue: :send_email_digest)
     end
   end
 
 private
 
-  def create_email(digest_run_subscriber, digest_items)
-    return if digest_items.empty?
-
+  def create_email(digest_run_subscriber, digest_item)
     Metrics.digest_email_generation(digest_run_subscriber.digest_run.range) do
       email = DigestEmailBuilder.call(
         address: digest_run_subscriber.subscriber.address,
-        digest_items: digest_items,
+        digest_item: digest_item,
         digest_run: digest_run_subscriber.digest_run,
         subscriber_id: digest_run_subscriber.subscriber_id,
       )
-      fill_subscription_content(email, digest_items, digest_run_subscriber)
 
-      email
+      email.id
     end
   end
 
-  def fill_subscription_content(email, digest_items, digest_run_subscriber)
+  def fill_subscription_content(email_ids, digest_items, digest_run_subscriber)
     now = Time.zone.now
-    records = digest_items.flat_map do |result|
-      result.content.map do |content|
+
+    records = digest_items.zip(email_ids).flat_map do |digest_item, email_id|
```

Comment:
There's quite a bit happening on this line and the next. My brain's struggling a bit to keep digest_items, records, and emails in play, probably because I'm not familiar with the code. Is there a way of pulling it apart a little that doesn't make it worse?

Reply:
I also stared at this for quite a while. Hopefully it's not much worse than it was before. I couldn't think of any meaningful variables to abstract things with - other than `digest_items_together_with_their_email`.

Fundamentally this is hard to understand because SubscriptionContent is hard to understand ([pain card for that](https://trello.com/c/0ge1rXr5/314-dev-pain-the-subscriptioncontent-model-is-confusing-and-hard-to-work-with)). But also, we push creating these join records away from creating the emails themselves i.e. new emails are initially orphaned. I think that's intentional ([see similarly here](https://github.com/alphagov/email-alert-api/blob/55726c8f73aabda0fd58588a6d3a942d3603ee86/app/services/immediate_email_generation_service/batch.rb#L15)) - i.e. new emails are temporarily orphaned from their content change / message / digest run - but does mean we end up with some heavy code to build the joining records manually, later on. Hopefully that makes some sense. I'm not sure there's much we can do with the code, though, short of looking again at why the SubscriptionContent model exists at all.

Reply:
A small amend you could make this easier to understand would be to insert into subscription contents multiple times as part of the [digest_items loop](https://github.com/alphagov/email-alert-api/pull/1524/files#diff-8e57211715641f37fe547d46d2a9665bc9e53503623cb9f7c1c598759cd6867bR16-R18). I can't imagine we benefit much from this being optimised to do only a single insert given we're already doing multiple email inserts.

Reply:
Oh yeah, that might be worth a try. I'll make a note to look at that in a refactoring PR.